From 890357f66e44f36474415148daf7eaa3532093cf Mon Sep 17 00:00:00 2001
From: Basheer Ahmed Muddebihal <basheer.ahmed.muddebihal@linux.intel.com>
Date: Fri, 27 Oct 2023 17:21:53 -0700
Subject: [PATCH] fpga: dfl: Add support for DFH Private feature type

Allow a Device Feature List (DFL) to start with a Device Feature Header
(DFH) of type Private feature in addition to the currently supported
type, Feature Management Entity (FME) or Port.

Signed-off-by: Basheer Ahmed Muddebihal <basheer.ahmed.muddebihal@linux.intel.com>
---
 drivers/fpga/dfl-pci.c | 5 +++++
 drivers/fpga/dfl.h     | 7 +++++++
 2 files changed, 12 insertions(+)

diff --git a/drivers/fpga/dfl-pci.c b/drivers/fpga/dfl-pci.c
index 7eb38a6a6bc1..347b00273604 100644
--- a/drivers/fpga/dfl-pci.c
+++ b/drivers/fpga/dfl-pci.c
@@ -313,6 +313,11 @@ static int find_dfls_by_default(struct pci_dev *pcidev,
 		start = pci_resource_start(pcidev, 0);
 		len = pci_resource_len(pcidev, 0);
 
+		dfl_fpga_enum_info_add_dfl(info, start, len);
+	} else if (dfl_feature_is_private(base)) {
+		start = pci_resource_start(pcidev, 0);
+		len = pci_resource_len(pcidev, 0);
+
 		dfl_fpga_enum_info_add_dfl(info, start, len);
 	} else {
 		v = readq(base + DFH);
diff --git a/drivers/fpga/dfl.h b/drivers/fpga/dfl.h
index ac6ed3b410fd..21f3c6736c35 100644
--- a/drivers/fpga/dfl.h
+++ b/drivers/fpga/dfl.h
@@ -496,6 +496,13 @@ static inline bool dfl_feature_is_port(void __iomem *base)
 		(FIELD_GET(DFH_ID, v) == DFH_ID_FIU_PORT);
 }
 
+static inline bool dfl_feature_is_private(void __iomem *base)
+{
+	u64 v = readq(base + DFH);
+
+	return (FIELD_GET(DFH_TYPE, v) == DFH_TYPE_PRIVATE);
+}
+
 static inline u8 dfl_feature_revision(void __iomem *base)
 {
 	return (u8)FIELD_GET(DFH_REVISION, readq(base + DFH));
